cloud/services/domains.go

```diff
@@ -0,0 +1,171 @@
+package services
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/linode/linodego"
+	kutil "sigs.k8s.io/cluster-api/util"
+
+	"github.com/linode/cluster-api-provider-linode/cloud/scope"
+)
+
+var dnsTTLSec = 30
+
+// AddIPToDNS creates domain record for machine public ip
+func AddIPToDNS(ctx context.Context, mscope *scope.MachineScope) error {
+	// Check if instance is a control plane node
+	if !kutil.IsControlPlaneMachine(mscope.Machine) {
+		return nil
+	}
+
+	// Get the public IP that was assigned
+	publicIP, err := GetMachinePublicIP(ctx, mscope)
+	if err != nil {
+		return fmt.Errorf("failed to get public IP of machine: %w", err)
+	}
+
+	// Get domainID from domain name
+	domainID, err := GetDomainID(ctx, mscope)
+	if err != nil {
+		return fmt.Errorf("failed to get domain ID: %w", err)
+	}
+
+	// Check if record exists for this IP and name combo
+	domainHostname := mscope.LinodeCluster.ObjectMeta.Name + "-" + mscope.LinodeCluster.Spec.Network.DNSUniqueIdentifier
+	filter, err := json.Marshal(map[string]interface{}{"name": domainHostname, "target": publicIP})
+	if err != nil {
+		return fmt.Errorf("failed to marshal domain filter: %w", err)
+	}
+
+	domainRecords, err := mscope.LinodeDomainsClient.ListDomainRecords(ctx, domainID, linodego.NewListOptions(0, string(filter)))
+	if err != nil {
+		return fmt.Errorf("unable to get current DNS record from API: %w", err)
+	}
+
+	if mscope.LinodeCluster.Spec.Network.DNSTTLSec != 0 {
+		dnsTTLSec = mscope.LinodeCluster.Spec.Network.DNSTTLSec
+	}
+
+	// If record doesnt exist, create it else update it
+	if domainRecords == nil {
+		recordReq := linodego.DomainRecordCreateOptions{
```

Comment:
do we want to use a `TXT` record similarly to external-dns to claim ownership of a address?

Reply:
So an A record that we currently do and a TXT record for the domain to add ownership?